File: src/wu/seal/jsontokotlin/KotlinMaker.java
package wu.seal.jsontokotlin;

import com.google.gson.*;
import org.jetbrains.annotations.NotNull;

import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;

/**
 * Created by seal.wu on 2017/8/21.
 */
public class KotlinMaker {

    private String className;
    private JsonElement inputElement;

    private Set<String> toBeAppend = new HashSet<String>();

    public KotlinMaker(String className, String inputText) {
        this.inputElement = new TargetJsonElement(inputText).getTargetJsonElementForGeneratingCode();
        this.className = className;
    }

    public KotlinMaker(String className, JsonElement jsonElement) {
        this.inputElement = new TargetJsonElement(jsonElement).getTargetJsonElementForGeneratingCode();
        this.className = className;
    }

    public String makeKotlinData() {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("\n");

        JsonElement jsonElement = inputElement;
        if (jsonElement.isJsonObject()) {
            appClassName(stringBuilder);
            appendJsonObject(stringBuilder, (JsonObject) jsonElement);
        } else {
            throw new IllegalArgumentException("不可能有其它情况");
        }

        int index = stringBuilder.lastIndexOf(",");
        if (index != -1) {
            stringBuilder.deleteCharAt(index);
        }
        stringBuilder.append(")");
        for (String append : toBeAppend) {
            stringBuilder.append("\n");
            stringBuilder.append(append);
        }
        return stringBuilder.toString();
    }

    private void appClassName(StringBuilder stringBuilder) {
        stringBuilder.append("data class ").append(className).append("(\n");
    }


    private void appendJsonObject(StringBuilder stringBuilder, JsonObject jsonObject) {
        for (Map.Entry<String, JsonElement> entry : jsonObject.entrySet()) {
            String property = entry.getKey();
            JsonElement jsonElementValue = entry.getValue();
            String type = "String";
            if (jsonElementValue.isJsonArray()) {
                type = getArrayType(property, (JsonArray) jsonElementValue);
                addProperty(stringBuilder, property, type, "");
            } else if (jsonElementValue.isJsonPrimitive()) {
                type = getPrimitiveType(jsonElementValue);
                addProperty(stringBuilder, property, type, jsonElementValue.getAsString());
            } else if (jsonElementValue.isJsonObject()) {
                type = getJsonObjectType(property, (JsonObject) jsonElementValue);
                addProperty(stringBuilder, property, type, "");
            } else if (jsonElementValue.isJsonNull()) {
                addProperty(stringBuilder, property, type, null);
            }
        }
    }

    @NotNull
    private String getJsonObjectType(String property, JsonObject jsonElementValue) {
        String type;
        type = property.subSequence(0, 1).toString().toUpperCase() + property.subSequence(1, property.length());
        type = KClassName.INSTANCE.getLegalClassName(type);
        toBeAppend.add(new KotlinMaker(type, jsonElementValue).makeKotlinData());
        return type;
    }

    @NotNull
    private String getArrayType(String property, JsonArray jsonElementValue) {
        String type = "List<String>";
        JsonArray jsonArray = jsonElementValue;

        Iterator<JsonElement> iterator = jsonArray.iterator();
        if (iterator.hasNext()) {
            JsonElement next = iterator.next();
            if (next.isJsonPrimitive()) {
                String subType = getPrimitiveType(next);
                type = "List" + "<" + subType + ">";

            } else if (next.isJsonObject()) {
                property = modifyPropertyForArrayObjType(property);
                String subType = getJsonObjectType(property, (JsonObject) next);
                type = "List" + "<" + subType + ">";
                /**
                 * 处理子类
                 */
                toBeAppend.add(new KotlinMaker(subType, next).makeKotlinData());
            } else if (next.isJsonArray()) {
                property = modifyPropertyForArrayObjType(property);
                String subType = getArrayType(property, (JsonArray) next);
                type = "List" + "<" + subType + ">";

            } else if (next.isJsonNull()) {
                type = "List" + "<" + "String" + ">";

            }
        } else {
            type = "List<Any>";

        }
        return type;
    }

    @NotNull
    private String modifyPropertyForArrayObjType(String property) {
        if (property.endsWith("ies")) {
            property = property.substring(0, property.length() - 3).concat("y");
        } else if (property.contains("List")) {
            int firstLatterAfterListIndex = property.lastIndexOf("List") + 4;
            if (property.length() > firstLatterAfterListIndex) {
                char c = property.charAt(firstLatterAfterListIndex);
                if (c >= 'A' && c <= 'Z') {
                    String pre = property.substring(0, property.lastIndexOf("List"));
                    String end = property.substring(firstLatterAfterListIndex, property.length());
                    property = pre.concat(end);
                }
            } else if (property.length() == firstLatterAfterListIndex) {
                property = property.substring(0, property.lastIndexOf("List"));
            }
        } else if (property.contains("list")) {
            if (property.indexOf("list") == 0) {
                String end = property.substring(5);
                String pre = (property.charAt(4) + "").toLowerCase();
                property = pre.concat(end);
            }
        } else if (property.endsWith("s")) {
            property = property.substring(0, property.length() - 1);
        }

        return property;
    }

    private void addProperty(StringBuilder stringBuilder, String property, String type, String value) {
        if (value == null) {
            value = "null";
        }
        stringBuilder.append(new KProperty(property, type, value).getPropertyStringBlock());
        stringBuilder.append("\n");
    }


    @NotNull
    private String getPrimitiveType(JsonElement next) {
        String subType = "String";
        JsonPrimitive asJsonPrimitive = next.getAsJsonPrimitive();
        if (asJsonPrimitive.isBoolean()) {
            subType = "Boolean";
        } else if (asJsonPrimitive.isNumber()) {
            if (asJsonPrimitive.getAsString().contains(".")) {
                subType = "Double";
            } else if (asJsonPrimitive.getAsLong() > Integer.MAX_VALUE) {
                subType = "Long";
            } else {
                subType = "Int";
            }
        } else if (asJsonPrimitive.isString()) {
            subType = "String";
        }
        return subType;
    }

}

File: src/wu/seal/jsontokotlin/KotlinMaker.kt
package wu.seal.jsontokotlin

import com.google.gson.*

import java.util.HashSet

/**
 * Kotlin code maker
 * Created by seal.wu on 2017/8/21.
 */
class KotlinMaker {

    private var className: String? = null
    private var inputElement: JsonElement? = null

    private val toBeAppend = HashSet<String>()

    constructor(className: String, inputText: String) {
        this.inputElement = TargetJsonElement(inputText).getTargetJsonElementForGeneratingCode()
        this.className = className
    }

    constructor(className: String, jsonElement: JsonElement) {
        this.inputElement = TargetJsonElement(jsonElement).getTargetJsonElementForGeneratingCode()
        this.className = className
    }

    fun makeKotlinData(): String {
        val stringBuilder = StringBuilder()
        stringBuilder.append("\n")

        val jsonElement = inputElement
        if (jsonElement!!.isJsonObject) {
            appClassName(stringBuilder)
            appendCodeMember(stringBuilder, jsonElement.asJsonObject)
        } else {
            throw IllegalArgumentException("UnSupport")
        }

        val index = stringBuilder.lastIndexOf(",")
        if (index != -1) {
            stringBuilder.deleteCharAt(index)
        }
        stringBuilder.append(")")
        for (append in toBeAppend) {
            stringBuilder.append("\n")
            stringBuilder.append(append)
        }
        return stringBuilder.toString()
    }

    private fun appClassName(stringBuilder: StringBuilder) {
        stringBuilder.append("data class ").append(className).append("(\n")
    }


    private fun appendCodeMember(stringBuilder: StringBuilder, jsonObject: JsonObject) {

        for ((property, jsonElementValue) in jsonObject.entrySet()) {

            var type = "String"
            if (jsonElementValue.isJsonArray) {

                type = getArrayType(property, jsonElementValue.asJsonArray)

                if (isExpectedJsonObjArrayType(jsonElementValue.asJsonArray)) {
                    toBeAppend.add(KotlinMaker(type, jsonElementValue.asJsonArray.first()).makeKotlinData())
                }
                addProperty(stringBuilder, property, type, "")

            } else if (jsonElementValue.isJsonPrimitive) {
                type = getPrimitiveType(jsonElementValue)
                addProperty(stringBuilder, property, type, jsonElementValue.asString)

            } else if (jsonElementValue.isJsonObject) {
                type = getJsonObjectType(property)
                toBeAppend.add(KotlinMaker(type, jsonElementValue).makeKotlinData())
                addProperty(stringBuilder, property, type, "")

            } else if (jsonElementValue.isJsonNull) {
                addProperty(stringBuilder, property, type, null)
            }
        }
    }


    private fun addProperty(stringBuilder: StringBuilder, property: String, type: String, value: String?) {
        var innerValue = value
        if (innerValue == null) {
            innerValue = "null"
        }
        stringBuilder.append(KProperty(property, type, innerValue).getPropertyStringBlock())
        stringBuilder.append("\n")
    }

}

fun main(args: Array<String>) {
    isTestModel = true
    val json1 = """{ "progr ammers": [
                { "firstName": "Brett", "lastName":"McLaughlin", "email": "aaaa" },
                { "firstName": "Jason", "lastName":"Hunter", "email": "bbbb" },
                { "firstName": "Elliotte", "lastName":"Harold", "email": "cccc" }
                ],
                "aut_hors": [
                { "firstName": "Isaac", "lastName": "Asimov", "genre": "science fiction" },
                { "firstName": "Tad", "lastName": "Williams", "genre": "fantasy" },
                { "firstName": "Frank", "lastName": "Peretti", "genre": "christian fiction" }
                ],
                "musicians": [
                { "firstName": "Eric", "lastName": "Clapton", "instrument": "guitar" },
                { "firstName": "Sergei", "lastName": "Rachmaninoff", "instrument": "piano" }
                ] } """

    val json2 = """ {"123menu": {
                "i d": [1,23,34],
                "value":[],
                "popup": {
                "m#@!$#%$#^%*^&(*)*(_)+{|}{:enu_item": [
                {"value": "New", "onclick": "CreateNewDoc()"},
                {"value": "Open", "onclick": "OpenDoc()"},
                {"value": "Close", "onclick": "CloseDoc()"}
                ]}
                }}"""

    println("json1 ====>\n${KotlinMaker("Class1", json1).makeKotlinData()}")
    println("json2 ====>\n${KotlinMaker("Class2", json2).makeKotlinData()}")

    TestConfig.targetJsonConvertLib = TargetJsonConverter.None
    TestConfig.isCommentOff = true
    TestConfig.isPropertiesVar = true

    println("===========================================Change to none json lib support========================================= ")

    println("json1 ====>\n${KotlinMaker("Class1", json1).makeKotlinData()}")
    println("json2 ====>\n${KotlinMaker("Class2", json2).makeKotlinData()}")

}